### PR TITLE
Implementation of Multi-thread in test framework

### PR DESCRIPTION
**Pull request by ono-max**
These are example of test(debug) time for Github workflow.

### Before
Finished in 379.300202619 seconds.

https://github.com/ruby/debug/runs/3179106902

### After()
Finished in 199.289802343 seconds.

https://github.com/ruby/debug/runs/3179484608

#### - I removed all instance variables from `utils.rb` and `assertions.rb` for thread-safe.
  - The all instance variables are stored as `Struct` class.
  - Queue instance is created for each threads.
#### - `test-unit` doesn't support multi thread (FYI, https://github.com/test-unit/test-unit/issues/204)
  - That's why I don't use `test-unit`'s assert method under multi threads. As the alternative plan, I implemented as follows.
    - If a test fail, the current thread will do nonlocal exits.
    -  If a test succeeds, assert method doesn't do nothing and `assert true` will be run.
  - I am concerned that the number of assert's is not accurate.

### CODE REVIEW

File: test/support/utils.rb [hunk 1]
```diff
@@ -32,26 +32,59 @@ def debuggee_backlog
...
+        end
+      else
+        debug_on_local boot_options, TestInfo.new(new_queue)
       end
+      assert true
     end
 
-    ASK_CMD = %w[quit q delete del kill undisplay]
+    def new_thread &block
+      Thread.new do
```
Comment (ko1) [inline test/support/utils.rb:67]: If thread terminates with an exception, `t.join` on line 55 will raise an exception.
It should rescue all exceptions and return it as a failure reason.

File: test/support/utils.rb [hunk 2]
```diff
@@ -32,26 +32,59 @@ def debuggee_backlog
...
+        Thread.current[:is_subthread] = true
+        catch(:fail) do
+          block.call
+        end
+      end
+    end
+
+    def multithreading?
+      Thread.current[:is_subthread]
```
Comment (ko1) [inline test/support/utils.rb:76]: `Thread` is general terminology and this case, you may want to know it is a worker thread for debugger testing. You should use more specific term.

File: test/support/utils.rb [hunk 3]
```diff
@@ -32,26 +32,59 @@ def debuggee_backlog
...
+    end
+
+    def new_queue
```
Comment (ko1) [inline test/support/utils.rb:79]: It is better to use more specific name.

File: test/support/utils.rb (hunk 2 above)
Comment (ono-max) [reply test/support/utils.rb:80]: I changed it to `multithreaded_test?`.

File: test/support/utils.rb (hunk 1 above)
Comment (ono-max) [reply test/support/utils.rb:72]: https://github.com/ruby/debug/pull/180/files#diff-ee14bf65f14927f7778a246f90c5cec0c581be3e2658c2a8a3726c59db5be3f9R61-R64

File: test/support/utils.rb (hunk 2 above)
Comment (ono-max) [reply test/support/utils.rb:81]: https://github.com/ruby/debug/pull/180/files#diff-ee14bf65f14927f7778a246f90c5cec0c581be3e2658c2a8a3726c59db5be3f9R80

File: test/support/utils.rb (hunk 3 above)
Comment (ono-max) [reply test/support/utils.rb]: https://github.com/ruby/debug/pull/180/files#diff-ee14bf65f14927f7778a246f90c5cec0c581be3e2658c2a8a3726c59db5be3f9R84

Comment (ono-max) [reply test/support/utils.rb]: https://github.com/ruby/debug/pull/180/files#diff-ee14bf65f14927f7778a246f90c5cec0c581be3e2658c2a8a3726c59db5be3f9R71

**Comment by ko1** [comment]
🎉

**Pull request merged**